base/subarray.jl

```diff
@@ -3,30 +3,78 @@
 typealias NonSliceIndex Union{Colon, AbstractVector}
 typealias ViewIndex Union{Int, NonSliceIndex}
 
-# LD is the last dimension up through which this object has efficient
-# linear indexing. If LD==length(I), then the object itself has efficient
-# linear indexing.
-immutable SubArray{T,N,P<:AbstractArray,I<:Tuple{Vararg{ViewIndex}},LD} <: AbstractArray{T,N}
+# L is true if the view itself supports fast linear indexing
+immutable SubArray{T,N,P,I,L} <: AbstractArray{T,N}
```

Comment:
I played with a few permutations here… this one seemed to be the easiest.  The actual computation on `I` is pretty complicated, and it is definitely nice to be able to dispatch on the linear indexability.

Another path I tried (unsuccessfully) was to support strided indexing into any fast parent array ([mb/lispystrided](https://github.com/JuliaLang/julia/compare/master...mbauman:mb/lispystrided)).  Eg., you would then be able to index `(2:5, :, :, :)` with just two indices very rapidly (instead of requiring the full dimensionality of `LinearSlow`).  In general, however, the extra work doesn't seem to be worth it… unless I'm missing something.